tools/src/test/java/org/apache/kafka/tools/PushHttpMetricsReporterTest.java

```diff
@@ -0,0 +1,333 @@
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.powermock.api.easymock.PowerMock.replayAll;
+import static org.powermock.api.easymock.PowerMock.verifyAll;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(PushHttpMetricsReporter.class)
+public class PushHttpMetricsReporterTest {
+
+    private static final URL URL;
+    static {
+        try {
+            URL = new URL("http://fake:80");
+        } catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    private PushHttpMetricsReporter reporter;
+    private Time time = new MockTime();
+    @MockStrict
+    private ScheduledExecutorService executor;
+    private Capture<Runnable> reportRunnable = EasyMock.newCapture();
+    @MockStrict
+    private HttpURLConnection httpReq;
+    @MockStrict
+    private OutputStream httpOut;
+    private Capture<byte[]> httpPayload = EasyMock.newCapture();
+    @MockStrict
+    private InputStream httpErr;
+
+    @Before
+    public void setUp() {
+        reporter = new PushHttpMetricsReporter(time, executor);
+        PowerMock.mockStatic(PushHttpMetricsReporter.class);
+    }
+
+    @Test
+    public void testConfigureClose() throws Exception {
+        expectConfigure();
+        expectClose();
+
+        replayAll();
+
+        configure();
+        reporter.close();
+
+        verifyAll();
+    }
+
+    @Test(expected = ConfigException.class)
+    public void testConfigureBadUrl() throws Exception {
+        Map<String, String> config = new HashMap<>();
+        config.put(PushHttpMetricsReporter.METRICS_URL_CONFIG, "malformed;url");
+        config.put(PushHttpMetricsReporter.METRICS_PERIOD_CONFIG, "5");
+        reporter.configure(config);
+    }
+
+    @Test(expected = ConfigException.class)
+    public void testConfigureMissingPeriod() throws Exception {
+        Map<String, String> config = new HashMap<>();
+        config.put(PushHttpMetricsReporter.METRICS_URL_CONFIG, URL.toString());
+        reporter.configure(config);
+    }
+
+    @Test
+    public void testNoMetrics() throws Exception {
+        expectConfigure();
+        expectRequest(200);
+        expectClose();
+
+        replayAll();
+
+        configure();
+        reportRunnable.getValue().run();
+        JsonNode payload = new ObjectMapper().readTree(httpPayload.getValue());
+        assertTrue(payload.isObject());
+
+        assertPayloadHasClientInfo(payload);
+
+        // Should contain an empty list of metrics, i.e. we report updates even if there are no metrics to report to
+        // indicate liveness
+        JsonNode metrics = payload.get("metrics");
+        assertTrue(metrics.isArray());
+        assertEquals(0, metrics.size());
+
+        reporter.close();
+
+        verifyAll();
+    }
+
+    // For error conditions, we expect them to come with a response body that we can read & log
```

Comment:
In this an the next test, I am not sure what exactly is being tested.

Reply:
It's just testing that when there is an error, we read the response. Mainly we want to verify that we are getting and logging the response when there is an error, but testing for the logging step is harder. Even the actual `readResponse` is mocked out because mocking the calls that `Scanner` makes seemed way messier, error prone, and possibly even JVM-specific.